Gems/Terrain/Code/Source/TerrainRenderer/TerrainDetailMaterialManager.cpp

```diff
@@ -764,20 +777,29 @@ namespace Terrain
                             pixels.at(index).m_material1 = aznumeric_cast<uint8_t>(materialId);
                             firstWeight = surfaceTagWeight.m_weight;
                             isFirstMaterial = false;
-                            static constexpr float MaxValueBeforeRounding = 254.5f / 255.0f;
-                            if (firstWeight >= MaxValueBeforeRounding)
-                            {
-                                break;
-                            }
                         }
                         else
                         {
                             // Second material is valid, weight is relative based on first material's weight.
-                            pixels.at(index).m_material2 = aznumeric_cast<uint8_t>(materialId);
-                            float totalWeight = firstWeight + surfaceTagWeight.m_weight;
-                            float blendWeight = 1.0f - (firstWeight / totalWeight);
-                            pixels.at(index).m_blend = aznumeric_cast<uint8_t>(AZStd::round(blendWeight * 255.0f));
-                            break;
+                            if (surfaceTagWeight.m_weight < 1.0 - firstWeight)
+                            {
+                                // If the second material is less weight than what the default material would be, use
+                                // the default material instead
+                                uint8_t defaultMaterial = region->m_defaultDetailMaterialId == InvalidDetailMaterailId ? m_passthroughMaterialId :
+                                    aznumeric_cast<uint8_t>(m_detailMaterials.GetData(region->m_defaultDetailMaterialId).m_detailMaterialBufferIndex);
+
+                                pixels.at(index).m_material2 = defaultMaterial;
+                                float blendWeight = 1.0f - AZStd::clamp<float>(firstWeight, 0.0f, 1.0f);
+                                pixels.at(index).m_blend = aznumeric_cast<uint8_t>(AZStd::round(blendWeight * 255.0f));
+                            }
+                            else
+                            {
+                                float totalWeight = firstWeight + surfaceTagWeight.m_weight;
+                                pixels.at(index).m_material2 = aznumeric_cast<uint8_t>(materialId);
+                                float blendWeight = 1.0f - (firstWeight / totalWeight);
```

Comment:
If there are more than 2 surface tags, anything after the 2nd tag is going to have the same first weight, but an increasing total weight, so blendWeight is going to go up and up for every subsequent tag. Is that intended? Or can there only ever be 2 or fewer surface tags (in which case maybe assert the size is <=2)?

The logic of this loop is a bit hard to track. It might be worth separating it into its own function, then creating a unit test that calls the function with different variations of what's inside surfacePoint.m_surfaceTags. That would both validate the logic and serve as a way of documenting in the code what the expected final blend weights would be for a given set of surface tags.

Reply:
I'm working on altering this to be a bit more clear, but for starters this will never use more than 2 weights - the for loop may skip some weights if there's no material associated with it, but only two weights with materials are ever used because that's all we support. It's totally ok for it to have lots of weights, we just care about the top two which have material associations.

I think the unit test idea is good, but I'd like to do that in a separate review once we're happy with the math. The tricky part with this is figuring out which two materials to use and at which weights given the top two materials + a default material, and what's "right" gets a bit subjective. There will always be some cases that don't look good. (like having a bunch of weights which all have similar values)